public/app/plugins/datasource/loki/datasource.ts

```diff
@@ -176,16 +185,36 @@ export class LokiDatasource extends DataSourceApi<LokiQuery, LokiOptions> {
       .map(target => this.prepareLiveTarget(target, options));
 
     for (const liveTarget of liveTargets) {
-      const subscription = webSocket(liveTarget.url)
+      liveTarget.subscription = webSocket(liveTarget.url)
         .pipe(
-          map((results: any[]) => {
-            const delta = this.processResult(results, liveTarget);
+          map((response: any) => {
+            const { refId } = liveTarget;
+
+            // What should we do with:
+            // response.dropped_entries?
+
+            const streams: LokiLogsStream[] = response.streams;
+            if (streams) {
+              for (const stream of streams) {
+                const buffer = getBufferedDataForLiveTarget(stream.labels, liveTarget);
+
+                // Add each line
+                for (const entry of stream.entries) {
+                  buffer.times.add(entry.ts || entry.timestamp);
+                  buffer.lines.add(entry.line);
+                }
+                buffer.frame.length = buffer.times.length;
```

Comment:
So not really sure about the buffering here. If I get the code, this just caps the data creating a rolling window but I do not think it actually buffers. I mean on every websocket message we pass the data through (the rolling window with max cap). With simple buffer I would expect not to get any data until the buffer is full (with some manual `flush` method). Or have a time based buffer that buffers for some time window (or some combination of that). Maybe the behaviour is what we actually want here (in that case I assume we will implement the time based buffering as before outside of DS), but then I think the naming is misleading.

Also how this correlates with having delta/full data? I mean this seems to be the `full` data case for streaming, just with a max limit. In that case the options to me seem more like:
- give me `full` frame streaming with limit = 1000 (basically what the current buffering method is)
- give me `delta` (basically just passing on all the messages), optionally with max limit on per websocket message basis. In that case what to do with the messages would be on the caller.